table/tables/index.go

```diff
@@ -176,6 +180,9 @@ func (c *index) Create(sctx sessionctx.Context, txn kv.Transaction, indexedValue
 
 	// save the key buffer to reuse.
 	writeBufs.IndexKeyBuf = key
+	c.initNeedRestoreData.Do(func() {
+		c.needRestoredData = NeedRestoredData(c.idxInfo.Columns, c.tblInfo.Columns)
```

Comment:
We may need to create index during upgrade in the future.

Reply:
If the index creator has no idea about whether `new collation` is enabled, how can it create?

Reply:
We can enable it at the beginning of `upgrade`.

Reply:
But we don't need to do it in this PR.

Reply:
If the `upgrade()` changes the schema of `mysql.tidb`, we will encounter the same problem.